### PR TITLE
[CALCITE-6481] Optimize 'VALUES...UNION ALL...VALUES' to a single 'VA…

### PR DESCRIPTION
**Pull request by NobiGo**
…LUES' the IN-list contains NULL and it is converted to VALUES

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java [hunk 1]
```diff
@@ -258,7 +258,7 @@ protected boolean needToCast(SqlValidatorScope scope, SqlNode node,
     RelDataType fromType = validator.deriveType(scope, node);
     // This depends on the fact that type validate happens before coercion.
     // We do not have inferred type for some node, i.e. LOCALTIME.
```
Comment (caicancai) [inline core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java:260]: It seems that judging the type of node, null and SqlTypeName.NUL are not equivalent in my opinion. Do we need to improve this part of the comment?

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java [hunk 2]
```diff
@@ -258,7 +258,7 @@ protected boolean needToCast(SqlValidatorScope scope, SqlNode node,
...
-    if (fromType == null) {
+    if (fromType == null || fromType.getSqlTypeName() == SqlTypeName.NULL) {
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java:261]: I don't see why this is right, and I don't see why it's needed for this optimization.
In general many tests will fail if you supply a NULL, but will pass if you cast the NULL to the correct type. So I suspect this is actually needed.

Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java:261]: @mihaibudiu Yes, I will check again. After this PR, some unit test behavior is unexpected.

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java (hunk 1 above)
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java:260]: This repair method is not comprehensive enough. I will check again.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 3]
```diff
@@ -491,6 +491,28 @@ private static String toSql(RelNode root, SqlDialect dialect,
     relFn(relFn).ok(expected);
   }
 
+  @Test void testSelectWhereIn2() {
```
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:494]: how are these tests related to the issue solved?

Comment (mihaibudiu) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:494]: I am guessing that the IN list is converted to a union and it's converted back to a values
Perhaps you can add a comment to say that

File: core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+import com.google.common.collect.ImmutableCollection;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import org.immutables.value.Value;
+
+import java.util.List;
+
+/**
+ * Planner rule that convert
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java:37]: converts

File: core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+ * {@link org.apache.calcite.rel.core.Union} with
+ * input only containing {@link org.apache.calcite.rel.core.Values}
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java:39]: inputs

File: core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+ * <p>becomes
+ *
+ * <blockquote><pre>{@code
+ * Values(tuples=[[{ 3, null }, { 7369, null }, { 1, 2 }]])
+ * }</pre></blockquote>
+ *
+ * <p>If (<code>all</code> = <code>false</code>),
+ * {@link org.apache.calcite.rel.core.Values} value will be set list.
+ *
+ * <p>If {@link org.apache.calcite.rel.core.Union} only one input, will be handled by
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java:60]: I am not sure this documents the current rule. It is rather some advice to use the other rule to handle this case.

File: core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    call.transformTo(values);
+  }
+
+  private static ImmutableCollection.Builder<ImmutableList<RexLiteral>> getImmutableListBuilder(
+      Boolean all, List<RelNode> relNodeList) {
+    ImmutableCollection.Builder<ImmutableList<RexLiteral>> tupleList;
+    if (all) {
+      tupleList = ImmutableList.builder();
+    } else {
+      tupleList = ImmutableSet.builder();
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java:101]: What is concerning here is that two VALUES may have the same "value" but not be "equal".
In particular, I am not sure about implicit casts.
Perhaps you can add some tests like `VALUE(5.0) UNION ALL VALUES(5)` or
`VALUES(CAST 5.0 AS INT) UNION ALL VALUES(5)`

File: core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java (hunk 6 above)
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java:60]: I will remove this document. Because UNION_REMOVE only can handle `all = false`. But this rule can handle both when its input is `Values`.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java (hunk 3 above)
Comment (NobiGo) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:494]: I had the same idea before. I added this test case because I want to know if RelBuilder generates the UNION when IN-List has CAST. But It doesn't.

File: core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java (hunk 7 above)
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java:101]: Maybe we will find another issue with `VALUE(5.0) UNION ALL VALUES(5)`. 
Here is a test case.
```
!use post
select *
from (values (5.0)) as t(x)
union
select *
from (values (5)) as t(y);
X
5
5.0
!ok

EnumerableUnion(all=[false])
  EnumerableValues(tuples=[[{ 5.0 }]])
  EnumerableValues(tuples=[[{ 5 }]])
!plan

select *
from (values (cast(5.0 as int))) as t(x)
union
select *
from (values (5)) as t(y);
X
5
!ok

EnumerableUnion(all=[false])
  EnumerableValues(tuples=[[{ 5 }]])
  EnumerableValues(tuples=[[{ 5 }]])
!plan
```
But in Postgresql. The SQL will return 5.0:
```
select *
from (values (5.0)) as t(x)
union
select *
from (values (5)) as t(y);
```

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java:101]: Calcite's pervasive use of implicit casts makes solving such problems much harder than expected. Other compilers have a stage which converts implicit casts into explicit casts in the program representation, which is then always unambiguously typed. I think SqlToRel should do that, but that seems like a very difficult task.

Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java:93]: I agree with your opinion. In Calcite, implicit cast transformations are in different stages according to different operations. In this PR, I have kept this result `5.0 5`. We can move to another issue to resolve this. Now I'm not sure how big the impact of the issue is.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java:93]: The problem is that your transformation will no longer produce an equivalent program. If the UNION removes duplicates, and after transformation you have just VALUES with duplicates, the two programs are no longer computing the same result.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java:93]: you could check to see if the type of all VALUES is the same, and then you can still apply the transformation as it is, otherwise you leave the UNION in the program. The transformation will probably be correct for UNION ALL.

Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/rel/rules/UnionToValuesRule.java:96]: I added type validation according to your suggestion, and it seems that the result is correct.

**Review by mihaibudiu — APPROVED** [review]
